src/MongoDB.Driver.Core/Core/Servers/Server.cs

```diff
@@ -147,13 +147,22 @@ public IChannelHandle GetChannel(CancellationToken cancellationToken)
                 connection.Open(CancellationToken.None);
                 return new ServerChannel(this, connection);
             }
-            catch
+            catch (Exception ex)
             {
+                if (ShouldClearConnectionPool(ex))
+                {
+                    _connectionPool.Clear();
```

Comment:
I checked the implementation of ExclusiveConnectionPool.Clear and it appears that calling Clear simply abandons all existing connections without closing them.

Open a separate ticket for ensuring that Clear closes the sockets.

If Clear becomes a blocking operation then we're going to need a ClearAsync version also.

Closing the connections in the background is dangerous because it creates a possible race conditions where connections are opened faster than they are closed leading to resource exhaustion. But the details should be worked out when implementing the new ticket.

Reply:
Ticket created!